pkg/cmd/crdconversionwebhook/converter/converter.go

```diff
@@ -0,0 +1,199 @@
+	"github.com/openshift/console-operator/pkg/api"
+)
+
+func convertCRD(object *unstructured.Unstructured, toVersion string) (*unstructured.Unstructured, metav1.Status) {
+	originalObject := object.DeepCopy()
+	fromVersion := object.GetAPIVersion()
+	convertedObject := &unstructured.Unstructured{}
+	var err error
+
+	if toVersion == fromVersion {
+		return nil, statusErrorWithMessage("conversion from a version to itself should not call the webhook: %s", toVersion)
+	}
+
+	switch object.GetAPIVersion() {
+	case "console.openshift.io/v1alpha1":
+		switch toVersion {
+		case "console.openshift.io/v1":
+			klog.Infof("converting %q object from 'console.openshift.io/v1alpha1' into 'console.openshift.io/v1'", originalObject.GetName())
+			convertedObject, err = convertPluginV1alpha1ToV1(originalObject)
+			if err != nil {
+				return nil, statusErrorWithMessage("error converting %q object from 'console.openshift.io/v1alpha1' into 'console.openshift.io/v1: %v'", originalObject.GetName(), err)
+			}
+		default:
+			return nil, statusErrorWithMessage("unexpected conversion version %q", toVersion)
+		}
+	case "console.openshift.io/v1":
+		switch toVersion {
+		case "console.openshift.io/v1alpha1":
+			klog.Infof("converting %q object from 'console.openshift.io/v1' into 'console.openshift.io/v1alpha1'", originalObject.GetName())
+
+			convertedObject, err = convertPluginV1ToV1alpha1(originalObject)
+			if err != nil {
+				return nil, statusErrorWithMessage("error converting %q object from 'console.openshift.io/v1' into 'console.openshift.io/v1alpha1: %v'", originalObject.GetName(), err)
+			}
+		default:
+			return nil, statusErrorWithMessage("unexpected conversion version %q", toVersion)
+		}
+	default:
+		return nil, statusErrorWithMessage("unexpected conversion version %q", fromVersion)
+	}
+
+	return convertedObject, statusSucceed()
+}
+
+func convertPluginV1ToV1alpha1(convertedObject *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	v1alpha1Plugin := &v1alpha1.ConsolePlugin{
+		TypeMeta: metav1.TypeMeta{Kind: "ConsolePlugin", APIVersion: "console.openshift.io/v1alpha1"},
+	}
+	v1Plugin := &v1.ConsolePlugin{}
+
+	// convert unstructured object to v1 ConsolePlugin
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(convertedObject.Object, v1Plugin)
+	if err != nil {
+		return nil, err
+	}
+
+	// metadata
+	v1alpha1Plugin.ObjectMeta = v1Plugin.ObjectMeta
+
+	// displayName
+	v1alpha1Plugin.Spec.DisplayName = v1Plugin.Spec.DisplayName
+
+	// i18n
+	if v1Plugin.Spec.I18n.LoadType == v1.Preload {
+		annotations := v1Plugin.GetAnnotations()
+		if annotations == nil {
+			annotations = map[string]string{
+				api.PluginI18nAnnotation: "true",
+			}
+		} else {
+			annotations[api.PluginI18nAnnotation] = "true"
+		}
+
+		v1alpha1Plugin.SetAnnotations(annotations)
+	}
+
+	// backend -> service
+	// we only support backend type 'Service' for now, so it's always true
+	if v1Plugin.Spec.Backend.Type == v1.Service {
```

Comment:
Should we return an error if we ever get an unrecognized type? I could see this happening if we add another type later, but forget to update the conversion webhook.

Same question for the proxy service type.

Reply:
Right, so this was originally part of the PR, but @deads2k [commented](https://github.com/openshift/console-operator/pull/670#discussion_r968783091) on it.

Reply:
Ack disregard